klippa/src/serialize.rs

```diff
@@ -351,4 +551,171 @@ mod test {
+            Some(&(0, 0))
+        );
+        assert_eq!(hash_table.get_with_hash(2, hash_2, &data, &obj_pool), None);
+        assert_eq!(hash_table.get_with_hash(3, hash_3, &data, &obj_pool), None);
+
+        hash_table.set_with_hash(2, hash_2, 2, &data, &obj_pool);
+        assert_eq!(
+            hash_table.get_with_hash(2, hash_2, &data, &obj_pool),
+            Some(&(2, 2))
+        );
+
+        hash_table.set_with_hash(3, hash_3, 3, &data, &obj_pool);
+        assert_eq!(
+            hash_table.get_with_hash(3, hash_3, &data, &obj_pool),
+            Some(&(3, 3))
+        );
+
+        // update obj_3 to have the same real links as obj_0
+        {
+            let obj_3 = obj_pool.get_obj_mut(3).unwrap();
+            let link = Link {
+                width: LinkWidth::Two,
+                is_signed: false,
+                whence: OffsetWhence::Head,
+                bias: 0,
+                position: 21,
+                objidx: Some(2),
+            };
+            obj_3.real_links.push(link);
+        }
+
+        let hash_3_new = hash_one_pool_idx(3, &data, &obj_pool);
+        assert_ne!(hash_3, hash_3_new);
+        // old entries found with old hash
+        assert_eq!(
+            hash_table.get_with_hash(3, hash_3, &data, &obj_pool),
+            Some(&(3, 3))
+        );
+
+        // test del
+        hash_table.del(3, hash_3, &data, &obj_pool);
+        assert_eq!(hash_table.get_with_hash(3, hash_3, &data, &obj_pool), None);
+
+        // duplicate obj_0 entry found with new hash
+        assert_eq!(
+            hash_table.get_with_hash(3, hash_3_new, &data, &obj_pool),
+            Some(&(0, 0))
+        );
+    }
+
+    #[test]
+    fn test_push_and_pop_pack() {
+        let mut s = Serializer::new(100);
+        let n = Uint24::new(80);
+        assert_eq!(s.embed(n), Ok(0));
+        assert_eq!(s.head, 3);
+
+        //single pop_pack()
+        assert_eq!(s.push(), Ok(()));
+        assert_eq!(s.embed(n), Ok(3));
+        assert_eq!(s.head, 6);
+        assert_eq!(s.tail, 100);
+        assert_eq!(s.pop_pack(true), Some(1));
+        assert_eq!(s.packed_map.hash_table.len(), 1);
+        assert_eq!(s.head, 3);
+        assert_eq!(s.tail, 97);
+
+        //test de-duplicate
+        assert_eq!(s.push(), Ok(()));
+        assert_eq!(s.embed(n), Ok(3));
+        assert_eq!(s.head, 6);
+        assert_eq!(s.pop_pack(true), Some(1));
+        assert_eq!(s.packed_map.hash_table.len(), 1);
+        //check to make sure head is rewinded
+        assert_eq!(s.head, 3);
+        assert_eq!(s.tail, 97);
+
+        assert_eq!(s.push(), Ok(()));
+        let n = UfWord::new(10);
+        assert_eq!(s.embed(n), Ok(3));
+        assert_eq!(s.head, 5);
+        assert_eq!(s.pop_pack(true), Some(2));
+        assert_eq!(s.packed_map.hash_table.len(), 2);
+        //check to make sure head is rewinded
+        assert_eq!(s.head, 3);
+        //check that tail is updated
+        assert_eq!(s.tail, 95);
+    }
```

Comment:
Some additional tests to add:
- pop_pack(false) when a duplicate object exists
- de-duplication with virtual links (ie. checking merge virtual links works, and that virtual links don't affect equality for de-duping).
- de-duplication on an object that has real links to test they are being included in hash and equality computations.

Reply:
Thanks! Adding de-duplication test with virtual links uncovered a typo in the code: I used real_links in the merge_virtual_links() API.